internal/workflows/workflows.go

```diff
@@ -0,0 +1,168 @@
+)
+
+// TODO: Write tasks for the following:
+// TODO: Determine how to handle support for readwritemany vs readwriteonce
+// TODO: Determine what to do if the volume is not large enough
+// TODO: Secure the volume so only the pod from the workflow and read and write to it.
+// TODO: Optimize the steps so we don't download and upload project files, dependencies and artifacts multiple times.
+// TODO: Get logs working with pipelines
+
+func GenPipelineWf(t *v1alpha1.Task) (*argo.WorkflowSpec, error) {
+	var wfSteps []argo.ParallelSteps
+	for _, s := range t.Spec.Pipeline {
+		tSpec := genTaskSpecFromPipelineStep(t, s)
+		wfTemplate := genTemplate(s.TaskName, v1alpha1.NewTaskFromSpec(tSpec, t.GetName()+s.TaskName), taskVolName(t))
+		wfSteps = append(wfSteps, argo.ParallelSteps{
+			Steps: []argo.WorkflowStep{
+				{
+					Inline: &wfTemplate,
+					Name:   s.TaskName,
+				},
+			},
+		})
+	}
+
+	volStoreageReq, err := v1resources.ParseQuantity("5Gi")
+	if err != nil {
+		return nil, err
+	}
+
+	return &argo.WorkflowSpec{
+		PodMetadata: &argo.Metadata{
+			Labels: map[string]string{
+				"ame-task": t.GetName(),
+			},
+		},
+		Templates: []argo.Template{
+			{
+				Steps: wfSteps,
+				Name:  "main",
+			},
+		},
+		Entrypoint: "main",
+		VolumeClaimTemplates: []apiv1.PersistentVolumeClaim{
+			{
+				ObjectMeta: v1.ObjectMeta{
+					Name: taskVolName(t),
+				},
+				Spec: apiv1.PersistentVolumeClaimSpec{
+					AccessModes: []apiv1.PersistentVolumeAccessMode{
+						apiv1.ReadWriteOnce,
+					},
+					Resources: apiv1.ResourceRequirements{
+						Requests: apiv1.ResourceList{
+							"storage": volStoreageReq,
+						},
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+func taskVolName(t *v1alpha1.Task) string {
+	return fmt.Sprintf("%s-volume", t.GetName())
+}
+
+func genTaskSpecFromPipelineStep(parentTask *v1alpha1.Task, step v1alpha1.PipelineStep) *v1alpha1.TaskSpec {
+	spec := parentTask.Spec.DeepCopy()
+	spec.RunCommand = step.RunCommand
+	spec.Pipeline = nil
+	spec.Env = step.Env
+	spec.Secrets = step.Secrets
+	return spec
+}
+
+func genTemplate(templateName string, t *v1alpha1.Task, volName string) argo.Template {
+	return argo.Template{
+		Name: templateName,
+
+		Script: &argo.ScriptTemplate{
+			Source: fmt.Sprintf(`
+
+          export TASK_DIRECTORY=ameprojectstorage/%s
+
+          s3cmd --no-ssl --region us-east-1 --host=$MINIO_URL --host-bucket=$MINIO_URL get --recursive s3://$TASK_DIRECTORY ./
+
+          cd "./%s" 
+
+          set -e # It is important that the workflow exits with an error code if execute or save_artifacts fails, so AME can take action based on that information.
+
+          execute "%s" 
+          
+          save_artifacts "ameprojectstorage/%s/artifacts/"
+
+          echo "0" >> exit.status
+					`, t.Spec.ProjectId, t.Spec.ProjectId, t.Spec.RunCommand, t.GetName()),
+			Container: apiv1.Container{
+				Image: "ame-executor:local",
+				Command: []string{
+					"bash",
+				},
+				Env: append([]apiv1.EnvVar{
```

Comment:
I think the template environmet construction be shard between this file and controllers/argo_workflows.go